### DISCUSSION TITLE
chap 9 질문 정리

### DISCUSSION THREAD

**Discussion by sihyung92** [JS-OS-study/hyokyung-and-students#33] [General]
가상메모리

**Comment by sihyung92** [comment]
# 느낀 점
- 가상 메모리 부분은 '이것을 쓰고 있다' 는게 명확하지 않아 잘 안 와닿는 파트였던 것 같다.
- 그래도 프로세스를 잘라서 따로 올린다는 페이징 기법과, '프로세스에서 자주 사용하는 부분은 작다' 는 발상에서 발전한 요구 페이징 기법은 발상이 정말 대단한것 같다.

# 질문
- LRU, LFU가 Page Replacement에서 활용하기 어려운 이유는 무엇일까요!?
- MPD, 즉 동시에 활용되는 프로세스 개수를 늘려주면 CPU 활용률은 계속해서 증대된다 (O / X)

**Reply by pullingoff** [reply]
1. 페이지 폴트가 일어났을 때에만 운영체제가 관여하기 때문에 가장 자주 쓰이는 페이지와 가장 최근에 참조된 페이지를 운영체제가 알 수 없기 때문
2. X. 프로세스가 스왑 인/아웃 하느라 CPU 활용률이 결국엔 떨어진다.

**Reply by 1chz** [reply]
- 페이지 폴트가 일어나면 인터럽트가 발생해 OS에게 CPU 제어권이 넘어가는데 이 시점에 운영체제는 어떤 페이지가 가장 오래되거나 많이 참조된 것인지 알기 힘들다
- X
  - CPU 사용률을 높이기 위해 `동시에 실행하는 프로세스의 수(MPD, Degree of Multiprogramming)`를 올린다. 왜냐하면 동시에 실행되는 프로세스의 수가 늘어나면, 어떤 프로세스가 I/O 작업으로 인해 `Blocked` 됐을 때 `컨텍스트 스위칭`을 통해 다른 프로세스를 작업하여 CPU가 노는 시간을 줄일 수 있기 때문이다.
  - 하지만 프로세스의 수를 늘리게 되면, 이 프로세스들을 위해 물리 메모리에 페이지 프레임을 할당해야 한다. 하지만 물리 메모리의 크기는 한정돼있기 때문에, 모든 프로세스의 모든 페이지 프레임을 물리 메모리에 올릴 수 없다. 따라서, 당장 실행해야만 하는 프로세스의 일부분만 물리 메모리에 올리게 되고, 이로 인해 페이지 폴트와 디스크 스왑 I/O가 필연적으로 많이 발생하게 된다
  - CPU가 노는 시간을 줄이기 위해 프로세스를 많이 올렸지만, 결과적으로 디스크 스왑 I/O가 더욱 빈번하게 발생하여 CPU 사용률이 오히려 더 떨어지게 된다
  - 이러한 현상을 `스레싱`이라 하며, `스레싱이 발생하는 근본적인 원인은 물리 메모리가 부족한 것`이다. 따라서, 스레싱의 해결 방안은 다음과 같다
    - 하드웨어적인 해결 방안: 물리 메모리를 더 늘려 스레싱이 발생하는 시기를 뒤로 밀어낸다 
    - 소프트웨어적인 해결 방안: 워킹셋을 사용한다
      - 한 프로세스가 실행 될 때 너무 많은 페이지 프레임을 할당 해버리면, 다른 프로세스가 사용할 페이지 프레임이 줄어든다(파이는 한정돼있다)
      - 반대로 너무 적은 페이지 프레임을 할당하면 빈번한 페이지 폴트가 발생해 디스크 스왑 I/O가 늘어나고 결국 스레싱이 발생한다
      - 프로세스가 실행될 때 일정량의 페이지 프레임을 할당하고, 만약 페이지 폴트가 발생하면 더 많은 페이지 프레임을 할당 해버린다. 그리고 페이지 폴트가 발생하지 않으면 오히려 페이지 프레임을 너무 많이 할당했다고 판단하여 이를 일부 회수한다
      - 이를 반복하면 해당 프로세스에게 맞는 적절한 페이지 프레임 수를 알아낼 수 있고, 적절한 페이지 프레임 수를 알아냈다면 해당 수 내에서 지역성 이론을 따라 많이, 자주 사용되는 페이지를 추출해 하나의 세트로 묶어 취급한다. 이를 `워킹셋`이라 부른다
  - 즉, 애초에 MPD가 높지 않아 물리 메모리가 부족하지 않다면 스레싱은 절대 발생하지 않는다. 따라서 이미 물리 메모리가 하는 작업을 충분히 감당하고 있는 상태라면, 물리 메모리를 더 추가한다고 해도 실제 성능 차이는 드라마틱하게 상승하지 않을 수 있다

**Reply by inseo24** [reply]
1. 페이지 폴트가 일어난 경우 cpu가 os로 넘어가는데 이 시점의 os는 가장 최근에 참조된 페이지를 알 수 없음
2. X, cpu 이용률 떨어짐

**Comment by inseo24** [comment]
<h3>느낀 점 & 궁금한 점 </h3>

- 맨날 지금 안 쓰는 거 먼저 알려주고 뒷통수를 치시는 거 같다. 
- 이번 챕터는 잘 이해되지 않는 것 투성이었다. 🤢

<hr />
<h3>문제</h3>

Q. CPU가 논리 주소를 갖고 메모리를 가져오려면 먼저 주소 변환을 하러 간다. 주소 변환시 invalid임을 확인하면 이 페이지가 메모리에 (있다/없다)라는 얘기다. 이 때는 해당 페이지를 디스크에서 메모리로 올려야 하는데 이는 i/o 작업으로 사용자 프로세스가 할 수 없는 일이다. 이 경우 CPU는 자동적으로 ____에게 넘어가게 된다. 이런 현상을 ___ _____라고 한다.

A. 
1. (있다/없다) 선택
2. ____(한글 4글자, 영어 답변 무관)
3. ____ _____(영어 9 words)

**Reply by pullingoff** [reply]
없다/운영체제/Page Fault

**Reply by 1chz** [reply]
1. 없다
2. OS
3. Page Fault

**Reply by sihyung92** [reply]
1. 없다
2. OS
3. page fault

**Comment by pullingoff** [comment]
# 느낀 점
 지금까지 운영체제 수업에서 배운 거의 모든 것들이 그랬듯이 페이지 교체 알고리즘들도 각각의 장단점, 트레이드오프가 있었다. 프로그래밍을 할 때는 정말x100 내가 가진 도구를 잘 알고, 그때그때 잘 써먹을 수 있게 계속 지식을 업데이트해두어야하는 것 같다.

# 문제
1. LRU 알고리즘의 한계점은? 
3. 멀티프로그래밍의 정도가 강해져 thrashing이 발생했을 때, 프로세스는 swap in/out을 하느라 바쁘다. 그러므로 CPU의 사용률이 매우 높은 상태이다. (O/X)
4. 워킹 셋의 크기(워킹셋 별 페이지의 수)는 각 운영체제마다 정해져있다. (O/X)

**Reply by 1chz** [reply]
1. 상동
2. X
3. O

**Reply by inseo24** [reply]
1. 최근은 아니더라도 비교적 자주 참조되는 페이지를 지울 수도 있음
2. x, 쓰레싱이 발생하면 cpu 사용률은 급격히 떨어지게 된다.(이용률은 쓰레싱이 발생하기 직전 최고치를 찍음)
3. x? working set window로 결정했던 거 같은데..

**Reply by sihyung92** [reply]
1. LFU와 마찬가지로 가장 최근에 참조되는 페이지에 대해 알 수 없다.
2. X
3. O, 윈도우 사이즈가 미리 정해져있지 않을까 라는 뇌피셜 발사

**Comment by 1chz** [comment]
# 느낀점

---

- JVM 레벨을 다루는 서적들에서 `페이지 폴트` 라는 용어가 굉장히 많이 나오지만 그 이유를 정확히 몰랐었는데, 이번에 그 이유를 어느 정도 알게 된 것 같다
- 마찬가지로 JVM 서적들에서 `가상 메모리` 어쩌고 하던 부분들도 어느 정도 이해가 되기 시작해서 좋았다
- **프로세스를 여러 개로 자른다**, **프로세스에서 자주 사용하는 부분은 작다** 같은 개념들이 운영체제 뿐만 아니고 다른 여러 분야에도 굉장히 많이 활용되는 것 같다고 생각했다
- 스레싱이 신기했다

<br />

---

![image](https://user-images.githubusercontent.com/71188307/158056458-271aa750-b3af-4d87-bfbe-e3a4de79b2d0.png)

<br />

---

![image](https://user-images.githubusercontent.com/71188307/158056406-917b65f8-0f2c-4b0a-a053-3998e386b70f.png)

<br />

---

![image](https://user-images.githubusercontent.com/71188307/158056240-7a48bb73-e8b7-4735-aee5-0def6646e2b1.png)

<br />


# 문제

---

- `페이지 폴트` 는 많이 일어날수록 `좋다 | 나쁘다`
- 이유가 무엇인가?

**Reply by inseo24** [reply]
나쁘다, OS에 cpu를 주고 받는 스위칭이 발생하기도 하고, 페이지 폴트가 발생한 스레드는 대기 상태가 되기도 하고.. 여러 모로 손실이 있다.

**Reply by pullingoff** [reply]
나쁘다. 처리하는데 비용이 많이 소모됨

**Reply by sihyung92** [reply]
나쁘다
메모리의 장점은 데이터 혹은 인스트럭션에 대한 빠른 접근인데, page fault가 발생하면 디스크에서 해당 내용을 찾아와야 한다(느린 접근).

**Comment by pullingoff** [comment]
swap id 로 어떤 프로세스가 스왑을 가장 많이하는지 확인하고 메모리 누수를 줄이거나 하드디스크나 메모리를 늘려야한다.
